main.py

```
@@ -1,240 +1,195 @@
+        else:
+            print("Tests failed. Build errors:\n" + build_result)
+        # Generate unit tests
+        past_actions = past_actions if 'past_actions' in locals() else []
+        max_retries = 3
+        retry_delay = 2  # seconds
+
+        for attempt in range(max_retries):
+            try:
+                unit_tests = refine_code_based_on_errors(
+                    sut=sut_content,
+                    test_cases=unit_tests_first if len(past_actions) == 0 else parsed.new_unit_test_code,
```

Ruff

`parsed` may be undefined, or defined from star imports

(F405)